primitives/src/rescue/sponge.rs

```diff
@@ -0,0 +1,441 @@
+    }
+
+    /// Pseudorandom function for Bls12_381 scalar field. It allows unrestricted
+    /// variable length input and number of output elements. Return error if
+    /// input is not multiple of STATE_SIZE = 4
+    pub fn full_state_keyed_sponge_no_padding(
+        key: &F,
+        input: &[F],
+        num_outputs: usize,
+    ) -> Result<Vec<F>, RescueError> {
+        assert_eq!(
+            CHUNK_SIZE, STATE_SIZE,
+            "CHUNK_SIZE must be equal to STATE_SIZE. Perhaps you meant to use a RescueCRH instead?"
+        );
+        if input.len() % STATE_SIZE != 0 {
+            return Err(RescueError::ParameterError(
+                "Rescue FSKS PRF Error: input to prf function is not multiple of STATE_SIZE."
+                    .to_string(),
+            ));
+        }
+        // ABSORB PHASE
+        let mut state = RescueVector::zero();
+        state.vec[STATE_SIZE - 1] = *key;
+        let mut r = Self::from_state(state, &Permutation::default());
+        r.absorb(&input);
+        Ok(r.squeeze_native_field_elements(num_outputs))
+    }
+}
+
+impl<F: RescueParameter, const CHUNK_SIZE: usize> SpongeExt for RescueSponge<F, CHUNK_SIZE> {
+    type State = RescueVector<F>;
+
+    fn from_state(state: Self::State, _permutation: &Self::Parameters) -> Self {
+        Self { state }
+    }
+
+    fn into_state(self) -> Self::State {
+        self.state
+    }
+}
+
+impl<T: RescueParameter + PrimeField, const CHUNK_SIZE: usize> CryptographicSponge
+    for RescueSponge<T, CHUNK_SIZE>
+{
+    /// Parameters used by the sponge.
+    type Parameters = Permutation<T>;
+
+    /// Initialize a new instance of the sponge.
+    fn new(_permutation: &Self::Parameters) -> Self {
+        Self {
+            state: RescueVector::default(),
+        }
+    }
+
+    /// Absorb an input into the sponge.
+    /// This function will absorb the entire input, in chunks of `CHUNK_SIZE`,
+    /// even if the input lenght is not a multiple of `CHUNK_SIZE`.
+    fn absorb(&mut self, input: &impl Absorb) {
+        let permutation = Permutation::default();
+
+        let input_field_elements = input.to_sponge_field_elements_as_vec();
+
+        // Absorb input.
+        input_field_elements
+            .chunks(CHUNK_SIZE)
+            .into_iter()
+            .for_each(|chunk| {
+                self.state.add_assign_elems(chunk);
+                self.state = permutation.eval(&self.state)
+            });
+    }
+
+    /// Squeeze `num_bytes` bytes from the sponge.
+    fn squeeze_bytes(&mut self, num_bytes: usize) -> Vec<u8> {
+        self.squeeze_bits(num_bytes)
+            .chunks(8)
+            .map(bools_to_u8)
+            .collect()
+    }
+
+    /// Squeeze `num_bits` bits from the sponge.
+    fn squeeze_bits(&mut self, num_bits: usize) -> Vec<bool> {
+        // we extract k number of field elements;
+        // each field elements will produce a maximum
+        // ```
+        // T::size_in_bits() - 129
```

Comment:
N/A due to https://github.com/EspressoSystems/jellyfish/pull/146/commits/8ef1f140705bac8ce639b934cb14576e82a45086.

I will add the comments later as a warning. Also, in the `unimplemented()` we can add an extra message, I'll do that too.

Reply:
Hi @tessico , let me know when the warning comments are done. I'll then approve.

Reply:
@chancharles92 I added the comments in https://github.com/EspressoSystems/jellyfish/pull/146/commits/2c817a4baf3b2fe9718ef03b3aaab18dbfaeca4b.
What I've also done in that commit was to completely "disable" the unintended usage of `CryptographicSponge`.
For our use case, we only want to allow users to squeeze **full native field elements**, meaning e.g. we shouldn't support truncated elements squeezing, as was previously done in e.g. `squeeze_native_field_elements_with_sizes`.

Under the hood, these methods would revert to `squeeze_native_field_elements` anyway if the `sizes` was full, and panic otherwise. The new changes reflect that under-the-hood behaviour, but are more explicit about what's supported and what's not, whereas previously a user might be tricked into using `squeeze_native_field_elements_with_sizes`, but then the method could panic if the input was not as expected.